src/Microsoft.Security.Utilities.Core/SecretMasker.cs

```diff
@@ -173,175 +176,124 @@ public string MaskSecrets(string input, Action<Detection>? detectionAction = nul
+                {
+                    _encodedSecretLiterals.Add(new SecretLiteral(encodedValue));
+                }
             }
         }
-
-        // Compute the encoded values.
-        var encodedSecrets = new List<SecretLiteral>();
-        foreach (SecretLiteral originalSecret in originalSecrets)
+        finally
         {
-            string encodedValue = encoder(originalSecret.Value);
-            if (!string.IsNullOrEmpty(encodedValue) && encodedValue.Length >= MinimumSecretLength)
-            {
-                encodedSecrets.Add(new SecretLiteral(encodedValue));
-            }
+            SyncObject.ExitWriteLock();
         }
+    }
 
-        // Write section.
+    public IEnumerable<Detection> DetectSecrets(string input)
+    {
+        var stopwatch = Stopwatch.StartNew();
         try
         {
-            SyncObject.EnterWriteLock();
-
-            // Add the encoder.
-            _ = LiteralEncoders.Add(encoder);
-
-            // Add the values.
-            foreach (SecretLiteral encodedSecret in encodedSecrets)
-            {
-                _ = EncodedSecretLiterals.Add(encodedSecret);
-            }
+            return DetectSecretsCore(input);
         }
         finally
         {
-            if (SyncObject.IsWriteLockHeld)
-            {
-                SyncObject.ExitWriteLock();
-            }
+            Interlocked.Add(ref _elapsedMaskingTicks, stopwatch.ElapsedTicks);
         }
     }
 
-    public IEnumerable<Detection> DetectSecrets(string input)
+    private List<Detection> DetectSecretsCore(string input)
     {
         if (string.IsNullOrEmpty(input))
         {
-            yield break;
+            return [];
         }
 
-        // Read section.
+        SyncObject.EnterReadLock();
         try
         {
-            SyncObject.EnterReadLock();
-            var stopwatch = Stopwatch.StartNew();
+            var detections = new List<Detection>();
+            List<HighPerformanceDetection>? highPerformanceDetections = _highPerformanceScanner?.Scan(input);
 
-            if (_highPerformanceScanner != null)
+            if (highPerformanceDetections != null)
             {
-                foreach (HighPerformanceDetection highPerformanceDetection in _highPerformanceScanner.Scan(input))
+                foreach (HighPerformanceDetection highPerformanceDetection in highPerformanceDetections)
                 {
                     string found = input.Substring(highPerformanceDetection.Start, highPerformanceDetection.Length);
                     foreach (RegexPattern pattern in _highPerformanceSignatureToPatternsMap![highPerformanceDetection.Signature])
                     {
                         Detection? detection = FinalizeHighPerformanceDetection(highPerformanceDetection, found, pattern);
-                        if (detection != null)
+                        if (detection != null && detection.Length >= MinimumSecretLength)
```

Comment:
>MinimumSecretLength [](http://example.com/codeflow?start=69&length=19)

See, this confuses me. I think the point of `MinimumSecretLength` is to allow short circuiting of detection and masking operations if your input data is so short that there can't be a secret there. It doesn't make sense to say, we found something, but it's too short to report.